awx/ui_next/Makefile

```diff
@@ -0,0 +1,97 @@
+## UI_NEXT_DIR: Relative path to the directory containing this Makefile
+UI_NEXT_DIR := $(patsubst %/,%,$(dir $(lastword $(MAKEFILE_LIST))))
+
+# ## Path to your local clone of the UI_NEXT repo
+# #  NOTE: you will not be able to build within the docker-compose development environment if you use this option
+UI_NEXT_LOCAL ?=
+
+# Git repo and branch to the UI_NEXT repo
+UI_NEXT_GIT_REPO ?= https://github.com/ansible/ansible-ui.git
+UI_NEXT_GIT_BRANCH ?= main
+
+.PHONY: ui-next
+## Default build target of ui-next Makefile, builds ui-next/build
+ui-next: ui-next/build
+
+.PHONY: ui-next/build
+## Build ui-next/build
+ui-next/build: $(UI_NEXT_DIR)/build
+
+## Real target for ui-next.
+$(UI_NEXT_DIR)/build: ui-next/src/build
+	@echo "=== Copying $(UI_NEXT_DIR)/src/build to $(UI_NEXT_DIR)/build ==="
+	@rm -rf $(UI_NEXT_DIR)/build
+	@cp -r $(UI_NEXT_DIR)/src/build $(UI_NEXT_DIR)
+	@echo "=== Done building $(UI_NEXT_DIR)/build ==="
+
+.PHONY: ui-next/src/build
+## Build ui-next/src/build
+ui-next/src/build: $(UI_NEXT_DIR)/src/build
+
+## Real target for ui-next/src/build.
+$(UI_NEXT_DIR)/src/build: ui-next/src ui-next/src/webpack
+	@echo "=== Building ui_next ==="
+	@cd $(UI_NEXT_DIR)/src && npm run build:awx
+
+.PHONY: ui-next/src
+ui-next/src: $(UI_NEXT_DIR)/src
+
+.PHONY: $(UI_NEXT_DIR)/src
+## Real target for ui-next/src.
+$(UI_NEXT_DIR)/src:
+	@echo "=== Setting up $(UI_NEXT_DIR)/src ==="
+	@if [ ! -z "$(UI_NEXT_LOCAL)" ]; then \
+		if [ -d $(UI_NEXT_DIR)/src ]; then \
+			if [ "$$(readlink $(UI_NEXT_DIR)/src)" = "$(UI_NEXT_LOCAL)" ]; then \
+				echo "SKIP: ui-next/src. $(UI_NEXT_DIR)/src already linked to $(UI_NEXT_LOCAL)."; \
+			else \
+				echo "=== Linking $(UI_NEXT_DIR)/src to $(UI_NEXT_LOCAL) ==="; \
+				rm -rf $(UI_NEXT_DIR)/src; \
+				ln -s $(UI_NEXT_LOCAL) $(UI_NEXT_DIR)/src; \
+			fi; \
+		else \
+			echo "=== Linking $(UI_NEXT_DIR)/src to $(UI_NEXT_LOCAL) ==="; \
+			ln -s $(UI_NEXT_LOCAL) $(UI_NEXT_DIR)/src; \
+		fi; \
+	elif [ ! -z "$(UI_NEXT_GIT_REPO)" ]; then \
+		if [ -d $(UI_NEXT_DIR)/src ]; then \
+			GIT_REMOTE_ORIGIN=$$(cd $(UI_NEXT_DIR)/src && git remote get-url origin); \
+			GIT_REMOTE_BRANCH=$$(cd $(UI_NEXT_DIR)/src && git rev-parse --abbrev-ref HEAD); \
+			if [ "$$GIT_REMOTE_ORIGIN" = "$(UI_NEXT_GIT_REPO)" ] && [ "$$GIT_REMOTE_BRANCH" = "$(UI_NEXT_GIT_BRANCH)" ]; then \
+				echo "=== Updating $(UI_NEXT_DIR)/src from $(UI_NEXT_GIT_BRANCH) of $(UI_NEXT_GIT_REPO) ==="; \
+				git fetch && git pull; \
+			else \
+				echo "=== Cloning $(UI_NEXT_DIR)/src from $(UI_NEXT_GIT_BRANCH) of $(UI_NEXT_GIT_REPO) ==="; \
```

Comment:
Again, echoing John's comment: this has warts. The first code smell is that it's just too long. In fact, that would cause `make` to fail in some environments because of argument list length restrictions! Make essentially passes everything as one long line as arguments to `sh -c`. Yes, it could use reworking, but even then it's probably doing too much. My first thought is just *lift*. Lift it to a helper script, possibly with params--you can always pass make macro vars `$()` into the helper script as-needed. Then, I don't even care that it's not great script, and it will be more likely to be improved later, will be picked up by shell linters automatically, etc.
That's a lie, tho. My first thought is really to let it in if it works well enough :)
But, if and when we want to fix it, we need to simplify it here and/or move it to a script.

Reply:
as discussed previously will address this outside of this PR...yes i hate this blob of bash too

Reply:
added TODO to address this later